Switch to wlr scene-graph

tiosgz opened the pull request:
~~It is still far from done, but apart from (not) raising view on focus it should be at least usable.~~

~~Depends on #44 (current commits from there are included) and wlroots 0.15 release.~~

TODOs –

- [ ] whatever’s a TODO in the code
- [x] use the scene as much as possible (e.g. cursor focus, though that may require further progress of wlroots too)
- [ ] WONTDO don’t make assumptions about new scene nodes (z-index, en/disabled…)
- [x] get rid of state duplication between the scene and kiwmi structs (positioning etc)
- [x] test it
- [ ] LATER maybe provide replacements for view pre/post-render – but it might be saner to keep for later

HeavyRain266 commented:
Definitely looks interesting, not sure what convinced them to use stuff from gamedev and vector graphics editors in 2d linux and bsd desktops though, Scene-graph is widely used practice in game engines and programs in style of Adobe Illustrator to easier view data structure, mayby it makes sense for them 🤔.

tiosgz commented:
6bd805cadd2283b4f702eaab6cc56c5713c9ba59 doesn’t fix the problem i thought it would fix (TODO: remove it, no need to move to another PR).

I’m not very motivated to write any code at all these days, but i hope i’ll get into it again soon enough.

Here’s a basic-tasks checklist i’ve managed to think of; pls let me know / edit if you come up with something more. Haven’t really tried to fill in the checks yet (again, i’m missing the motivation).

- scene
  - [x] created and destroyed
  - [x] bound to output layout
- [strata](https://github.com/tiosgz/kiwmi/blob/6bd805cadd2283b4f702eaab6cc56c5713c9ba59/include/desktop/desktop.h#L29-L32)
  - [ ] created
  - [ ] correctly sorted
  - [x] destroyed (automatically on scene destroy)
- output
  - [ ] nodes created and destroyed
  - [ ] nodes moved when moving
- view
  - [ ] node created and destroyed
  - [ ] surface geom updated in scene (and it must be correct)
  - [ ] node moved when moving
  - [ ] unmaps when unmapped / hidden
  - [ ] subsurfaces handled correctly (should be done by wlroots)
- layer
  - [ ] node created and destroyed
  - [ ] node moved when moving
  - [ ] node reparented on layer change
  - [ ] unmaps when unmapped
- popup
  - [ ] node created
  - [ ] node destroyed by wlroots?
  - [ ] node moved when moving (handled by wlroots)
  - [ ] correctly unconstrained
- bg color
  - [ ] node (rect) created and destroyed
  - [ ] node resized (and moved?) according to output layout
  - [ ] changed on user request (kiwmi:bg_color)
- not required for merge (IMO)
  - [ ] an alternative to view:pre/post_render

Question for myself: is output scaling correctly applied? (should be a wlr thing)

tiosgz commented on kiwmi/desktop/desktop.c:
```diff
@@ -43,6 +65,39 @@ desktop_init(struct kiwmi_desktop *desktop, struct wlr_renderer *renderer)
+        return false;
+    }
+
+    // different ordering:
+    // - background, bottom, views, top
+    // - bg+bot+views+top popups (possibly unified)
+    // - overlay, overlay popups
```
Any comments on this pls? I believe it’s the ‘correct’ order, but am unsure if the four popup strata should be reduced to a single stratum instead.

```diff
- current state
+ proposal
  --------
  background
  bottom
  views
  top
- background popups
- bottom popups
- view popups
- top popups
+ popups (except overlay)
  overlay
  overlay popups
```

tiosgz replied:
Hmm. I think i’ll keep them separated, as that is only a lil bit more complicated to initialize, but then easy to work with. Unless, of course, it becomes a single node with overlay’s popups.

tiosgz commented on include/desktop/toplevel.h:
```diff
@@ -0,0 +1,46 @@
+
+// prolly has to be initialized by the specific type of toplevel?
+struct kiwmi_toplevel {
+    struct wlr_scene_tree *scene_subtree;
+    struct wlr_scene_tree *popups_scene_subtree;
+    // in the future, scene_subtree will contain more than just this
+    struct wlr_scene_node *surface_scene_node;
```
I seem to have issues finding the sweet point between clarity & brevity. Better names?

tiosgz commented on include/desktop/output.h:
```diff
@@ -9,22 +9,23 @@
     struct wl_listener mode;
 
     struct wl_list layers[4]; // struct kiwmi_layer_surface::link
+    struct wlr_scene_tree *layer_strata_subtrees[4];
+    struct wlr_scene_tree *layer_popups_strata_subtrees[4];
```
Ditto.

tiosgz commented on kiwmi/desktop/view.c:
```diff
@@ -84,42 +84,26 @@ view_set_size(struct kiwmi_view *view, uint32_t width, uint32_t height)
 }
 
 void
+// why does it have to be unsigned? and why can it be negative, then?
 view_set_pos(struct kiwmi_view *view, uint32_t x, uint32_t y)
```
Unrelated, but this has been a mystery for me.

tiosgz commented on kiwmi/input/cursor.c:
```diff
@@ -284,65 +284,15 @@ cursor_refresh_focus(
-            &sx,
-            &sy);
-    }
+    // TODO: erm, shouldn't we ask for any surface instead? /shrug
+    struct kiwmi_toplevel *toplevel = toplevel_at(
```
I’m not quite sure about this one. If it’s about focusing a surface, should we really only ask for toplevels? (Although there isn’t any surface that couldn’t get us to a toplevel atm.)

Commit:
Remove API that relies on our rendering

This is a preparation for switching to the wlroots scene-graph. There is
no replacement planned for output:redraw() because it will no longer be
necessary (as far as i can tell), while scene-compatible replacements
for the view pre_render/post_render events are planned.

Commit:
Keep an up-to-date wlr scene-graph

At the moment it isn't used at all and a lot of data is duplicated. This
is a temporary state introduced for nicer diff separation into several
commits.

Commit:
Support protocols made easy by wlr_scene

tiosgz commented:
Cleaned up the commit history so that it should now be easier to understand. However, there are still things to be done (see checklist in PR description).

tiosgz commented on include/desktop/desktop.h:
```diff
@@ -9,24 +9,36 @@
+    struct wlr_scene_tree *layer_strata[4];
+    struct wlr_scene_tree *layer_popups_strata[4];
+    struct wlr_scene_tree *views_stratum;
+    struct wlr_scene_tree *view_popups_stratum;
```
Going with that (see f31f0e1a591dd9ad90f53693e592283d93be6489). Lemme know if you have some problem with it.

tiosgz commented on kiwmi/desktop/view.c:
```diff
@@ -247,29 +189,37 @@ view_resize(struct kiwmi_view *view, uint32_t edges)
- */
-void
-view_init_subsurfaces(struct kiwmi_view_child *child, struct kiwmi_view *view)
+static struct kiwmi_output *
+view_toplevel_get_output(struct kiwmi_toplevel *toplevel)
```
Kinda self-note: should try using `wlr_scene_surface::primary_output`, though that may be somewhat hacky as we don’t have direct access to the `wlr_scene_surface` (we could also use an algorithm similar to theirs instead).

tiosgz replied:
Not gonna do this, actually.

tiosgz commented on kiwmi/desktop/layer_shell.c:
```diff
@@ -12,26 +12,65 @@
 
-    wlr_layer_surface_v1_close(layer->layer_surface);
-
-    arrange_layers(layer->output);
```
self-note: this can apparently be of use when waybar is ctrl-c'd (with this removed, the exclusive zone didn’t disappear)

tiosgz commented on include/desktop/toplevel.h:
```diff
@@ -0,0 +1,44 @@
+struct wlr_surface;
+struct kiwmi_desktop;
+
+enum kiwmi_toplevel_type {
+    KIWMI_TOPLEVEL_VIEW,
+    KIWMI_TOPLEVEL_LAYER,
+};
+
+struct kiwmi_toplevel {
```
Apparently, *toplevel* is more widely used to refer only to xdg-shell or xwayland, but not layer-shell. Should think of a name that doesn’t conflict with other compositors’ terminology. (Perhaps i have accidentally swapped *toplevel* and *view*? 🤷)

HeavyRain266 replied:
Toplevel is the term used for windows, layershell represents viewports of surfaces.

tiosgz replied:
Self-note in case me forgets before me gets to renaming it: weston uses `desktop_surface` (pointed out to me by ammen99)

Commit:
Introduce kiwmi_strata

With the migration to wlr_scene, some objects (desktop, output etc)
will no longer have any reason to care about individual shells and
similar; handling each one on its own would introduce code that is,
from their perspective, over-complicated.

Strata (sg. stratum; this name was chosen to avoid confusion with the
over-used 'layer') represent logical layers of surfaces on the desktop,
such as every layer shell layer, one for xdg shell, another for popups.

tiosgz commented:
This force-push is mostly renaming stuff and moving commits around iirc; there should be no real changes.

tiosgz commented:
I’m planning only one change (plus the two minor TODOs somewhere in the code), namely getting rid of `view::hidden` and `view::mapped`. The PR is ready for review, and quite soon also for merge.

Commit:
Render using wlr_scene

Apart from removing a lot of code, this commit also introduces (thanks
to wlr_scene) fine-grained damage tracking.

The lack of our own rendering code will only be an issue if the need for
custom effects (such as blur) arises; however, that's very unlikely and
it would probably require a lot of changes anyway.

Commit:
Prefer wlr_scene as the source of truth

Use it for getting the surface under cursor, position of surfaces, etc.

Commit:
Move popup unconstraining logic

Commit:
Cleanup unneeded code after switching to wlr_scene

Commit:
(Un)focus logic improvements

tiosgz commented:
This force-push is mostly splitting commits; only minimal changes.

tiosgz commented on include/desktop/stratum.h:
```diff
@@ -0,0 +1,33 @@
+    KIWMI_STRATUM_LS_BACKGROUND, // LS == layer_shell
+    KIWMI_STRATUM_LS_BOTTOM,
+    KIWMI_STRATUM_NORMAL,
+    KIWMI_STRATUM_LS_TOP,
+    KIWMI_STRATUM_LS_OVERLAY,
+    KIWMI_STRATUM_POPUPS,
```
Other compositors have *top, normal popups, fulscreen, overlay, overlay popups*, but since kiwmi doesn’t support client-initiated fulscreen, i left it like this.

(This is just a note for later ig.)

Commit:
output: While destroying, destroy layers

Also handle NULL output in output_damage().

Co-authored-by: Uks2 <tom@tomg.xyz>

tiosgz commented on include/desktop/desktop_surface.h:
```diff
@@ -0,0 +1,45 @@
+void desktop_surface_get_pos(
+    struct kiwmi_desktop_surface *desktop_surface,
+    int *x,
+    int *y);
```
Whoops, fixed.

buffet commented:
Thank you very much!
